engine/flink/management/periodic/src/main/scala/pl/touk/nussknacker/engine/management/periodic/Utils.scala

```diff
@@ -1,13 +1,72 @@
 package pl.touk.nussknacker.engine.management.periodic
 
+import akka.actor.{ActorNotFound, ActorPath, ActorRef, ActorSystem}
+import akka.pattern.{AskTimeoutException, gracefulStop}
 import com.typesafe.scalalogging.LazyLogging
 
+import java.util.concurrent.TimeoutException
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{Await, Future}
+
 object Utils extends LazyLogging {
 
+  private val GracefulStopTimeout         = 5 seconds
+  private val ActorResolutionTimeout      = 5 seconds
+  private val TotalActorResolutionTimeout = 16 seconds
+
   def runSafely(action: => Unit): Unit = try {
     action
   } catch {
     case t: Throwable => logger.error("Error occurred, but skipping it", t)
   }
 
+  def gracefulStopActor(actorRef: ActorRef, actorSystem: ActorSystem): Unit = {
+    logger.info(s"Gracefully stopping $actorRef")
+
+    try {
+      Await.result(
+        gracefulStop(actorRef, GracefulStopTimeout),
+        GracefulStopTimeout + (1 second)
+      )
+    } catch {
+      case _: AskTimeoutException | _: TimeoutException =>
+        throw new IllegalStateException(s"Failed to gracefully stop actor $actorRef within timeout")
+    }
+
+    logger.info(s"Gracefully stopped $actorRef, waiting for it's name to be freed")
+
+    try {
+      Await.result(
+        waitUntilActorNameIsFree(actorRef.path, actorSystem),
```

Comment:
We do you have to wait until name is free if actor already answer on GracefulStopTimeout within configured time?

Reply:
Actor termination doesn't mean that `ActorSystem` is instantly aware of it: https://doc.akka.io/japi/akka/current/akka/pattern/GracefulStopSupport.html

>  **IMPORTANT NOTICE:** the actor being terminated and its supervisor being informed of the availability of the deceased actor&rsquo;s name are two distinct operations, which do not obey any reliable ordering. Especially the following will NOT work:

Reply:
LOL actor based programming is hard. I would add a short comment about that

Reply:
`// actor programming is hard`